### PR TITLE
Implements read_addrs

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**

### CODE REVIEW

File: gui/src/vmm/cpu/mod.rs [hunk 1]
```diff
@@ -256,16 +256,6 @@ impl<H: Hypervisor, S: Screen> CpuManager<H, S> {
         cpu: &mut impl Cpu,
         stop: Option<MultiThreadStopReason<u64>>,
     ) -> bool {
-        // Get states.
```
Comment (SuchAFuriousDeath) [inline gui/src/vmm/cpu/mod.rs:259]: I had to move this because of the borrow checker.

File: gui/src/vmm/cpu/debug.rs [hunk 2]
```diff
@@ -35,12 +35,21 @@ impl Debuggee {
...
         self.locked = true;
-        self.receiver
-            .recv()
-            .map(|v| match v {
-                DebugRes::Regs(v) => v,
-            })
-            .ok()
+        self.receiver.recv().ok().and_then(|v| match v {
+            DebugRes::Regs(v) => Some(v),
+            _ => None,
```
Comment (ultimaweapon) [inline gui/src/vmm/cpu/debug.rs:40]: This should be a panic.

File: gui/src/vmm/cpu/debug.rs [hunk 3]
```diff
@@ -35,12 +35,21 @@ impl Debuggee {
...
+    }
+
+    pub fn translate_address(&mut self, addr: usize) -> Option<usize> {
+        self.sender.send(DebugReq::TranslateAddress(addr)).ok()?;
+
+        self.locked = true;
+
+        self.receiver.recv().ok().and_then(|v| match v {
+            DebugRes::TranslatedAddress(v) => Some(v),
+            _ => None,
```
Comment (ultimaweapon) [inline gui/src/vmm/cpu/debug.rs:51]: Same here.

File: gui/src/vmm/cpu/mod.rs [hunk 4]
```diff
@@ -285,10 +275,31 @@ impl<H: Hypervisor, S: Screen> CpuManager<H, S> {
...
+                            unsafe { args.event.invoke(VmmEvent::Error { reason: &e }) };
+                            return false;
+                        }
+                    }
+                }
+                DebugReq::TranslateAddress(addr) => match cpu.translate(addr) {
+                    Ok(v) => debug.send(DebugRes::TranslatedAddress(v)),
                     Err(e) => {
-                        unsafe { args.event.invoke(VmmEvent::Error { reason: &e }) };
+                        let err = RustError::with_source("couldn't translate address", e);
```
Comment (ultimaweapon) [inline gui/src/vmm/cpu/mod.rs:300]: Better to include the address in the message.

File: gui/src/vmm/cpu/x86_64.rs [hunk 5]
```diff
@@ -66,11 +66,31 @@ impl<H: Hypervisor, S: Screen> MultiThreadBase for CpuManager<H, S> {
...
+            .ok_or(GdbTargetError::Errno(ENOENT))?;
+
+        drop(cpu);
 
         let ram = self.hv.ram();
 
-        todo!()
+        let locked_addr = ram
+            .lock(translated, len)
+            .ok_or(GdbTargetError::Errno(ENOENT))?;
```
Comment (ultimaweapon) [inline gui/src/vmm/cpu/x86_64.rs:87]: This should be `EFAULT`.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**